tracer/src/Datadog.Trace/Iast/ReturnedHeadersAnalyzer.cs

```diff
@@ -39,6 +41,127 @@ internal static void Analyze(IHeaderDictionary responseHeaders, IntegrationId in
+
+#if NETFRAMEWORK
+            foreach (var headerKey in responseHeaders.AllKeys)
+#else
+            foreach (var headerKey in responseHeaders.Keys)
+#endif
+            {
+                if (string.IsNullOrWhiteSpace(headerKey))
+                {
+                    continue;
+                }
+
+                bool isHeaderInjectionException = false;
+                foreach (var excludeType in headerInjectionExceptions)
+                {
+                    if (excludeType.Equals((string)headerKey, StringComparison.OrdinalIgnoreCase))
+                    {
+                        isHeaderInjectionException = true;
+                        continue;
+                    }
+                }
+
+                if (isHeaderInjectionException)
+                {
+                    continue;
+                }
+
+                string headerValue = responseHeaders[headerKey];
+
+                if (string.IsNullOrWhiteSpace(headerValue))
+                {
+                    continue;
+                }
+
+                var taintedValue = traceContext?.IastRequestContext?.GetTainted(headerValue);
+
+                if (taintedValue is null || taintedValue.Ranges.Length == 0)
+                {
+                    continue;
+                }
+
+                if (headerKey.Equals("set-cookie", StringComparison.OrdinalIgnoreCase))
+                {
+                    if (OnlyCookieValueSources(taintedValue))
+                    {
+                        continue;
+                    }
+                }
+
+                if (headerKey.Equals("access-control-allow-origin", StringComparison.OrdinalIgnoreCase))
+                {
+                    if (ComesFromOriginHeader(taintedValue))
+                    {
+                        continue;
+                    }
+                }
+
+                if (IsPropagationHeader(headerKey, taintedValue))
+                {
+                    continue;
+                }
+
+                IastModule.OnHeaderInjection(integrationId, headerKey, headerValue);
+            }
+        }
+        catch (Exception error)
+        {
+            Log.Error(error, $"Error in {nameof(ReturnedHeadersAnalyzer)}.{nameof(AnalyzeHeaderInjectionVulnerability)}");
+        }
+    }
+
+    private static bool OnlyCookieValueSources(TaintedObject taintedValue)
+    {
+        foreach (var range in taintedValue.Ranges)
+        {
+            if (range.Source is not null && range.Source.OriginByte != (byte)SourceTypeName.CookieValue)
+            {
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    private static bool IsPropagationHeader(string headerName, TaintedObject taintedValue)
+    {
+        return (
+            taintedValue.Ranges.Length == 1 &&
+            taintedValue.Ranges[0].Source is not null &&
+            taintedValue.Ranges[0].Source!.OriginByte == (byte)SourceTypeName.RequestHeaderValue &&
```

Comment:
I think OriginByte is not needed (it is not serialized) and makes code reading more difficult. It should be a SourceType directly

Reply:
Do you mean comparing the string value instead of the byte? This would be more expensive in terms of performance than just comparing a single byte. The byte value should always be present even if we don't serialize it.